src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -1454,6 +1454,99 @@ void MacroAssembler::update_word_crc32(Register crc, Register v, Register tmp1,
   xorr(crc, crc, tmp2);
 }
 
+void MacroAssembler::vector_update_crc32(Register crc, Register buf, Register len, const int64_t unroll_words,
+                                         Register tmp1, Register tmp2, Register tmp3, Register tmp4,
+                                         Register table0, Register table3) {
+
+    const int N = 16, W = 4;
+    const Register blks = tmp2;
+    const Register tmpTable = tmp3, tableN16 = tmp4;
+    const VectorRegister vcrc = v4, vword = v8, vtmp = v12;
+    Label VectorLoop;
+    Label LastBlock;
+
+    // prepare
+    add(tableN16, table3, 1*256*sizeof(juint), tmp1);
+    addi(len, len, unroll_words);
+
+    if (MaxVectorSize == 16) {
+      vsetivli(zr, N, Assembler::e32, Assembler::m4, Assembler::ma, Assembler::ta);
+    } else if (MaxVectorSize == 32) {
+      vsetivli(zr, N, Assembler::e32, Assembler::m2, Assembler::ma, Assembler::ta);
+    } else {
+      assert(MaxVectorSize > 32);
+      vsetivli(zr, N, Assembler::e32, Assembler::m1, Assembler::ma, Assembler::ta);
+    }
+    vmv_v_x(vcrc, zr);
+    slli(crc, crc, 32);
+    srli(crc, crc, 32);
+    vmv_s_x(vcrc, crc);
+
+    srli(blks, len, 6);
+    slli(t1, blks, 6);
+    sub(len, len, t1);
+    sub(blks, blks, 1);
+    blez(blks, LastBlock);
+
+    bind(VectorLoop);
+    {
+      mv(tmpTable, tableN16);
+
+      vle32_v(vword, buf);
+      vxor_vv(vword, vword, vcrc);
+
+      addi(buf, buf, N*W);
+
+      mv(t1, 0xff);
+      vand_vx(vtmp, vword, t1);
+      vsll_vi(vtmp, vtmp, 2);
+      vluxei32_v(vcrc, tmpTable, vtmp);
+
+      mv(tmp1, 1);
+      for (int k = 1; k < W; k++) {
+        addi(tmpTable, tmpTable, 256*4);
+
+        slli(t1, tmp1, 3);
+        vsrl_vx(vtmp, vword, t1);
+
+        mv(t1, 0xff);
+        vand_vx(vtmp, vtmp, t1);
+        vsll_vi(vtmp, vtmp, 2);
+        vluxei32_v(vtmp, tmpTable, vtmp);
+
+        vxor_vv(vcrc, vcrc, vtmp);
+
+        addi(tmp1, tmp1, 1);
+      }
+
+      sub(blks, blks, 1);
+      bgtz(blks, VectorLoop);
+    }
+
+    bind(LastBlock);
+    {
+      mv(crc, zr);
+      for (int i = 0; i < N; i++) {
+        lwu(t1, Address(buf, i*W));
+        vmv_x_s(t0, vcrc);
+        // in vmv_x_s, the value is sign-extended to SEW bits, but we need zero-extended here.
+        slli(t0, t0, 32);
```

Comment:
Could you use add.uw here to zero extend?  You appear to be using zba below in the shadd statement (I assuming this generates sh2add).